sources/web/datalab/settings.ts

```diff
@@ -172,33 +223,45 @@ function ensureDirExists(fullPath: string): boolean {
  */
 export function updateUserSetting(userId: string, key: string, value: string,
                                   asynchronous: boolean = false): boolean {
-  var userDir = userManager.getUserDir(userId);
-  var settingsDir =  path.join(userDir, 'datalab', '.config');
+  var settingsDir = getUserConfigDir(userId);
   var settingsPath = path.join(settingsDir, SETTINGS_FILE);
 
+  // Start by ensuring the containing directory exists; if we can't do that,
+  // we can't do anything else.
+  try {
+    if (!ensureDirExists(path.normalize(settingsDir))) {
+      console.log('Can\'t create containing directory for user settings, update not done.');
+      return false;
+    }
+  }
+  catch (e) {
+    console.log('Error creating directory for user settings', e);
+    return false;
+  }
+
   if (!fs.existsSync(settingsPath)) {
     console.log('User settings file %s not found, copying default settings.', settingsPath);
     copyDefaultUserSettings(userId);
   }
+  if (!fs.existsSync(settingsPath)) {
+    console.log('User settings file still doesn\'t exist, update not done.');
+    return false;
+  }
 
   let settings: common.UserSettings;
-  if (fs.existsSync(settingsPath)) {
-    try {
-      settings = <common.UserSettings>JSON.parse(fs.readFileSync(settingsPath, 'utf8') || '{}');
-    }
-    catch (e) {
-      console.log(e);
-      return false;
-    }
+  try {
+    settings = <common.UserSettings>JSON.parse(fs.readFileSync(settingsPath, 'utf8') || '{}');
```

Comment:
It looks like this code path was trying to avoid synchronous read/write operations. Can we turn this whole path into a promise (or async/await) chain?